File: examples/european_call/black_scholes.py
from scipy.stats import norm
from math import log, sqrt, exp

# European call option Black-Scholes formula

def black_scholes_call(S, K, T, r, sigma):
    d1 = (log(S / K) + (r + 0.5*sigma**2)*T) / (sigma * sqrt(T))
    d2 = d1 - sigma * sqrt(T)
    return S * norm.cdf(d1) - K * exp(-r*T) * norm.cdf(d2)

if __name__ == '__main__':
    # Match parameters in call_option.mlir and config.yaml
    S = 100.0
    K = 105.0
    T = 1.0
    r = 0.05
    sigma = 0.2
    price = black_scholes_call(S, K, T, r, sigma)
    print(f"Analytical Black-Scholes price: {price:.4f}")

File: runtime/black_scholes.cpp
//===- black_scholes.cpp - Black-Scholes Pricing ----------------*- C++ -*-===//
//
//  Black-Scholes formula for European option pricing.
//===----------------------------------------------------------------------===//

#include <cmath>
#include <algorithm>

namespace derivlab {

double norm_cdf(double x) {
    return 0.5 * std::erfc(-x * M_SQRT1_2);
}

double black_scholes_call(double S, double K, double r, double sigma, double T) {
    double d1 = (std::log(S / K) + (r + 0.5 * sigma * sigma) * T) / (sigma * std::sqrt(T));
    double d2 = d1 - sigma * std::sqrt(T);
    return S * norm_cdf(d1) - K * std::exp(-r * T) * norm_cdf(d2);
}

double black_scholes_put(double S, double K, double r, double sigma, double T) {
    double d1 = (std::log(S / K) + (r + 0.5 * sigma * sigma) * T) / (sigma * std::sqrt(T));
    double d2 = d1 - sigma * std::sqrt(T);
    return K * std::exp(-r * T) * norm_cdf(-d2) - S * norm_cdf(-d1);
}

} // namespace derivlab
